alacritty/src/display/damage.rs

```diff
@@ -0,0 +1,345 @@
+            // intersection as a result for current iteration.
+            if intersected_ratio >= MIN_DAMAGED_SQUARE_PART {
+                rect = intersected_rect;
+
+                // Reset look ahead state.
+                is_looking_ahead = false;
+                ahead_rect = rect;
+                ahead_square += next_rect_square;
+
+                // We've just added this rect into the intersected one, so the next one is unknown
+                // yet.
+                next_rect = None;
+
+                // Move index until we reset damaged lines and the one we'll use for next iteration.
+                next_index = index;
+            } else if intersected_ratio > MIN_DAMAGED_SQUARE_PART_ERR {
+                // The ratio increased by adding more rectangles to it, so continue getting
+                // rectangles.
+                ahead_rect = intersected_rect;
+                ahead_square += next_rect_square;
+
+                // In case we're starting looking ahead store information about the rectangle we'll
+                // be restoring to.
+                if !is_looking_ahead {
+                    next_rect = Some(ahead_next_rect);
+                    next_index = index;
+                }
+
+                // Start looking ahead.
+                is_looking_ahead = true;
+            } else {
+                break;
+            }
+
+            // Advance to the next undamaged rect.
+            match self.terminal_damage.iter().skip(index).position(|&l| l.is_damaged()) {
+                Some(relative_index) => {
+                    index += relative_index;
+                    ahead_next_rect =
+                        Self::rect_for_line(size_info, index, self.terminal_damage[index]);
+
+                    // If we're not looking ahead update next iteration information.
+                    if !is_looking_ahead {
+                        next_rect = Some(ahead_next_rect);
+                        next_index = index;
+                    }
+                },
+                None => {
+                    // If we're not looking ahead and reached the end make sure to reset bounds on
+                    // the rest of the lines.
+                    if !is_looking_ahead {
+                        next_index = self.terminal_damage.len();
+                    }
+
+                    break;
+                },
+            }
+        }
+
+        for index in self.index..next_index {
+            self.terminal_damage[index].reset();
+        }
+
+        // Update information for the next iteration.
+        self.rect = next_rect;
+        self.index = next_index;
+
+        Some(self.overdamage(rect))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use crate::config::UiConfig;
+    use alacritty_terminal::grid::Dimensions;
+    use alacritty_terminal::term::{
+        DamagedLine, RenderableCursor as TermRenderableCursor, SizeInfo, Term,
+    };
+
+    #[test]
+    fn damage_fully_damaged() {
```

Comment:
This module is already called `damage` we don't need to prefix every test method with `damage_`?